src/setPath.test.ts

```diff
@@ -0,0 +1,355 @@
+        },
+      });
+    });
+
+    test('should correctly type path argument', () => {
+      // @ts-expect-error - 'hello' isn't a valid path
+      expect<SampleType>(setPath(obj, ['a', 'hello'], 'hello')).toEqual({
+        ...obj,
+        a: {
+          ...obj.a,
+          hello: 'hello',
+        },
+      });
+    });
+
+    test('should work with undefined / optional types', () => {
+      expect<SampleType>(setPath(obj, ['a', 'z'], undefined)).toEqual({
+        ...obj,
+        a: {
+          ...obj.a,
+          z: undefined,
+        },
+      });
+    });
+
+    test('should support partial paths', () => {
+      expect<SampleType>(setPath(obj, ['a', 'b'], { c: 2 })).toEqual({
+        ...obj,
+        a: {
+          ...obj.a,
+          b: { c: 2 },
+        },
+      });
+    });
+
+    test('should correctly type partial paths', () => {
+      // @ts-expect-error - this path should yield a type of { c: number }
+      expect<SampleType>(setPath(obj, ['a', 'b'], 123)).toEqual({
+        ...obj,
+        a: {
+          ...obj.a,
+          b: 123,
+        },
+      });
+    });
+  });
+
+  describe('string path', () => {
+    test('should set a deeply nested value', () => {
+      expect<SampleType>(setPath(obj, 'a.b.c', 2)).toEqual({
+        ...obj,
+        a: {
+          ...obj.a,
+          b: {
+            c: 2,
+          },
+        },
+      });
+    });
+
+    test('should work nested arrays', () => {
+      expect<SampleType>(setPath(obj, 'a.e[1].f.g', 2)).toEqual({
+        ...obj,
+        a: {
+          ...obj.a,
+          e: [{ f: { g: 1 } }, { f: { g: 2 } }],
+        },
+      });
+    });
+
+    test('should correctly type value argument', () => {
+      expect<SampleType>(
+        // @ts-expect-error - this path should yield a type of number
+        setPath(obj, 'a.e[1].f.g', 'hello')
+      ).toEqual({
+        ...obj,
+        a: {
+          ...obj.a,
+          e: [{ f: { g: 1 } }, { f: { g: 'hello' } }],
+        },
+      });
+    });
+
+    test('should correctly type path argument', () => {
+      // @ts-expect-error - 'hello' isn't a valid path
+      expect<SampleType>(setPath(obj, 'a.hello', 'hello')).toEqual({
+        ...obj,
+        a: {
+          ...obj.a,
+          hello: 'hello',
+        },
+      });
+    });
+
+    test('should work with undefined / optional types', () => {
+      expect<SampleType>(setPath(obj, 'a.z', undefined)).toEqual({
+        ...obj,
+        a: {
+          ...obj.a,
+          z: undefined,
+        },
+      });
+    });
+
+    test('should support partial paths', () => {
+      expect<SampleType>(setPath(obj, 'a.b', { c: 2 })).toEqual({
+        ...obj,
+        a: {
+          ...obj.a,
+          b: { c: 2 },
+        },
+      });
+    });
+
+    test('should correctly type partial paths', () => {
+      // @ts-expect-error - this path should yield a type of { c: number }
+      expect<SampleType>(setPath(obj, 'a.b', 123)).toEqual({
```

Comment:
I'm not sure if we should support the dot notation. I don't see much advantage of it. What if we have an object that has a dot in the name, like?

```
const obj = {
 a: {
  b: 5
 },
 'a.b': 'five'
}
```

Reply:
I'm also hoping that the types might get less complex if we remove this, because honestly, I feel a bit intimidated by them 😅

Reply:
For me, the string notation is the primary reason I need a helper like this (migrating [this](https://www.remix-validated-form.io/arrays-and-nested) away from lodash is my use-case).

A lot of the type complexity also comes from supporting an arbitrary path length rather than only handling 3 like `pathOr` does. If we remove the string notation, it will simplify the types a little, but it will still be more complex than the types for `pathOr`.

What if we split this into two helpers, `setPath` and `stringToPath`? Then the string path use-case could be served this way:

```ts
setPath(obj, stringToPath('a.b.c'))
```

and then I can split up the types a bit more & add more comments to hopefully make it easier to grok.

Reply:
Also re: the object with `'a.b'` as a key:

My first instinct is to not handle that case, but I looked around at some of the prior art for this. The solution that makes the most sense to me is adding support for strings in the path like this: `a['b.c']`.

Makes things quite a bit more complex though. Almost makes me wonder if this kind of thing is worth having a dedicated library instead of adding it to a helper collection.

Reply:
> What if we split this into two helpers, setPath and stringToPath? Then the string path use-case could be served this way:

I personally like that suggestion a lot. We could also document the limitation that objects where the key itself contains a dot will be tricky to work with `stringToPath` and decide to not support that. People can always decide to _not_ use that helper and just use `setPath` with the array of path notation.

I think this means this helper would automatically be usable for `pathOr`:

```
pathOr(obj, stringToPath('a.b.c')
```

and we could simply / streamline the types of `pathOr` as well.

Gonna pull in @vlad-yakovlev to see what he thinks about the suggestion :)

Reply:
When I think about usability, I would like to split this functions in two. Like this: `setPath(obj, stringToPath('a.b.c'))`. But types looks impossible to me in this case.

So, if we need this feature and one day we need to use `a.b` key, we can use array argument for that

Reply:
Look like we need to pass `obj` to `stringToPath` to determine which keys we can use. If it's possible, then we definitely need to split this function in two

Reply:
Happy new year 🎉
no worries - I've been absent a bit as well :)